### PR TITLE
feat(application): Application DDL schema

### PR DESCRIPTION
**Pull request by SimonRichardson**
Follows on from [Charm SQL DDL Schema](https://github.com/juju/juju/pull/17508) change set.

----

Finishes off the application DDL schema. The application DDL schema already existed in a skeleton form, so the change set completes this work:

 - Adds password_hash_algorithm table to capture the passwords for machines and applications. Storing the algorithm allows us to migrate in the future.

 - Reworked the order of the files so that any foreign keys only reference files before them. A test will ensure that there are no cyclic foreign keys exist in the future.

 - scale and scale-target are on the application_caas. I'm not 100% sure about the name, and we iterate on that when we implement the service/state.

 - The channel is now stored as a separate table for applications and charms, so risk is a requirement.

----

As a driveby. Whilst adding application DDL and investigating what can and can not be a PRIMARY KEY, it was noticed that:

    > According to the SQL standard, PRIMARY KEY should always imply
    > NOT NULL. Unfortunately, due to a bug in some early versions,
    > this is not the case in SQLite.
    > ...
    > SQLite allows NULL values in a PRIMARY KEY column.

This commit fixes that to ensure they can't be NULL.

  1. https://www.sqlite.org/lang_createtable.html

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made

## QA steps

Tests should pass.

## Links

**Jira card:** [JUJU-6008](https://warthogs.atlassian.net/browse/JUJU-6008)

[JUJU-6008]: https://warthogs.atlassian.net/browse/JUJU-6008?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by wallyworld — APPROVED** [review]
move scale to application_caas and rename endpoints to endpoint

File: domain/schema/model/sql/0003-password.sql [hunk 1]
```diff
@@ -0,0 +1,10 @@
+CREATE TABLE password_hash_algorithm (
+    id INT PRIMARY KEY,
```
Comment (manadart) [inline domain/schema/model/sql/0003-password.sql:2]: Missed a `NOT NULL` here.

File: domain/schema/model/sql/0018-application.sql [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    application_uuid TEXT NOT NULL PRIMARY KEY,
+    track TEXT NOT NULL,
+    risk TEXT NOT NULL,
+    branch TEXT,
+    CONSTRAINT fk_application_channel_application
+    FOREIGN KEY (application_uuid)
+    REFERENCES application (uuid)
+);
+
+CREATE TABLE application_caas (
```
Comment (manadart) [inline domain/schema/model/sql/0018-application.sql:26]: I'd call this `application_scale`, and include a comment that it is only relevant to K8s.

File: domain/schema/model/sql/0018-application.sql [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    REFERENCES application (uuid),
+    CONSTRAINT fk_application_platform_os
+    FOREIGN KEY (os_id)
+    REFERENCES os (id),
+    CONSTRAINT fk_application_platform_architecture
+    FOREIGN KEY (architecture_id)
+    REFERENCES architecture (id)
+);
+
+CREATE UNIQUE INDEX idx_application_name
```
Comment (manadart) [inline domain/schema/model/sql/0018-application.sql:52]: This should live proximate to the application table.

File: domain/schema/model/sql/0018-application.sql (hunk 2 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0018-application.sql:26]: Much better, I'll swing back around and fix.

File: domain/schema/model/sql/0003-password.sql (hunk 1 above)
Comment (SimonRichardson) [reply domain/schema/model/sql/0003-password.sql:2]: Indeed

Comment (SimonRichardson) [reply domain/schema/model/sql/0003-password.sql:2]: No, int and integer doesn't need not null, only text does.

**Pull request merged**